keras_hub/src/models/qwen3_moe/qwen3_moe_attention.py

```diff
@@ -0,0 +1,376 @@
+import math
+
+import keras
+from keras import ops
+
+from keras_hub.src.layers.modeling.rotary_embedding import RotaryEmbedding
+from keras_hub.src.models.qwen3_moe.qwen3_moe_layernorm import Qwen3MoeLayerNorm
+from keras_hub.src.utils.keras_utils import clone_initializer
+from keras_hub.src.utils.keras_utils import fused_attention_op_available
+
+
+class Qwen3MoeAttention(keras.layers.Layer):
+    """A multi-head attention layer for Qwen3Moe models
+    This attention implementation supports grouped-query attention (GQA) where
+    the number of key-value heads can be less than the number of query heads.
+    Args:
+        num_query_heads: Number of query heads.
+        num_key_value_heads: Number of key/value heads (for GQA).
+        rope_max_wavelength: Maximum wavelength for RoPE (Rotary Position
+            Embedding).
+        rope_scaling_factor: Scaling factor for RoPE, used for extending
+            context length.
+        kernel_initializer: Initializer for the kernel weights.
+        dropout: Dropout rate for attention weights.
+        use_sliding_window_attention: Whether to use sliding window
+            attention.
+        sliding_window_size: Size of the sliding window for attention.
+        **kwargs: Additional keyword arguments to pass to the Layer.
+    """
+
+    def __init__(
+        self,
+        num_query_heads,
+        num_key_value_heads,
+        layer_index,
```

Comment:
This layer index is gross, let's remove it. Handle the args properly in the backbone and pass the correct `sliding_window_size` to this layer and the decoder layer above it.

Reply:
since it's an Moe, layer index is not just used for sliding window but for experts

Reply:
I replaced this passing of `layer_index`, `decoder_sparse_step` and `mlp_only_layers` with a single boolean switch:

https://github.com/kanpuriyanawab/keras-nlp/blob/730a9c41e95a74906a041d8933b8d7738391b438/keras_hub/src/models/qwen3_moe/qwen3_moe_backbone.py#L129-L156